earthfile2llb/converter.go

```diff
@@ -674,8 +693,14 @@ func (c *Converter) SaveArtifact(ctx context.Context, saveFrom string, saveTo st
 	if saveAsLocalTo != "" && c.opt.DoSaves {
 		separateArtifactsState := llbutil.ScratchWithPlatform()
 		if isPush {
+			pushState := persistCache(
+				c.mts.Final.RunPush.State,
+				c.persistentCacheDirs,
+				c.runOpts,
+				c.mts.Final.Platform,
+			)
 			separateArtifactsState = llbutil.CopyOp(
-				c.mts.Final.RunPush.State, []string{saveFrom}, separateArtifactsState,
+				pushState, []string{saveFrom}, separateArtifactsState,
```

Comment:
I thought it would be necessary to copy the cached files into the state used here as well. I've been testing more thoroughly though, and I don't think I understand when this case gets executed.

Can you or @alexcb explain a bit more? I understand we may set `isPush` to `true` by using `--push` flag in RUN` or `SAVE IMAGE`, but what is it's relation to `SAVE ARTIFACT`?

Reply:
Ah yes, I missed the part above containing

```go
			pushState := persistCache(
				c.mts.Final.RunPush.State,
				c.persistentCacheDirs,
				c.runOpts,
				c.mts.Final.Platform,
			)
```

I think this makes sense now. This is the case of

```earthly
RUN --push ...
SAVE ARTIFACT ... AS LOCAL ...
```

The `RUN --push` only executes during the push phase of the build, and therefore `SAVE ARTIFACT` would need to execute off of that in this case. So this makes sense now that I look at it more carefully.